logrotate.c

```diff
@@ -1388,6 +1388,98 @@ static long daysElapsed(const struct tm *now, const struct tm *last)
     return (long) ((intmax_t)diff / DAY_SECONDS);
 }
 
+static int isSecureDir(const char *pathname, uid_t exempt_uid, gid_t exempt_gid, int depth);
+
+static int isSecureDirStep(const char *pathname, uid_t exempt_uid, gid_t exempt_gid, int depth)
+{
+    struct stat sb;
+
+    if (lstat(pathname, &sb) < 0)
+        return -errno;
+
+    if (S_ISLNK(sb.st_mode)) {
+        /* check whether symlink target is inside a secure directory */
+        char *canonpath;
+        int r;
+
+        canonpath = realpath(pathname, NULL);
+        if (!canonpath)
+            return -errno;
+
+        r = isSecureDir(canonpath, exempt_uid, exempt_gid, depth + 1);
+        free(canonpath);
+        return r;
+    }
+
+    if (!S_ISDIR(sb.st_mode))
+        return -ENOTDIR;
+
+    if ((sb.st_uid != ROOT_UID && sb.st_uid != exempt_uid) ||
+        ((sb.st_mode & S_IWGRP) && sb.st_gid != 0 && sb.st_gid != exempt_gid) ||
+        (sb.st_mode & S_IWOTH && !(sb.st_mode & S_ISVTX))) {
```

Comment:
Is it safe to expect `S_ISVTX` to be defined on all systems where logrotate can be built?

Reply:
It's an XSI option, see https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/sys_stat.h.html.
Added a fallback definition.